authorizerd.go

```diff
@@ -152,13 +168,65 @@ func New(opts ...Option) (Authorizerd, error) {
 		jwkProvider = prov.jwkd.GetProvider()
 	}
 
-	prov.roleProcessor = role.New(
+	if prov.roleProcessor, err = role.New(
 		role.WithPubkeyProvider(pubkeyProvider),
-		role.WithJWKProvider(jwkProvider))
+		role.WithJWKProvider(jwkProvider),
+		role.WithEnableMTLSCertificateBoundAccessToken(prov.atpParams[0].verifyCertThumbprint),
+		role.WithClientCertificateGoBackSeconds(prov.atpParams[0].certBackdateDur),
+		role.WithClientCertificateOffsetSeconds(prov.atpParams[0].certOffsetDur),
+	); err != nil {
+		return nil, errors.Wrap(err, "error create role processor")
+	}
+
+	// create verifiers
+	if err = prov.initVerifiers(); err != nil {
+		return nil, errors.Wrap(err, "error create verifiers")
+	}
 
 	return prov, nil
 }
 
+func (a *authorizer) initVerifiers() error {
+	// TODO: check empty credentials to speed up the checking
+	verifiers := make([]verifier, 0, len(a.atpParams)+1+1)
+
+	if a.verifyRoleCert {
+		rcVerifier := func(r *http.Request, act, res string) error {
+			return a.VerifyRoleCert(r.Context(), r.TLS.PeerCertificates, act, res)
+		}
+		glg.Info("initVerifiers: added role certificate verifier")
+		verifiers = append(verifiers, rcVerifier)
+	}
+
+	if len(a.atpParams) > 0 {
+		atVerifier := func(r *http.Request, act, res string) error {
+			tokenString, err := request.AuthorizationHeaderExtractor.ExtractToken(r)
+			if err != nil {
+				return err
+			}
+			return a.VerifyAccessToken(r.Context(), tokenString, act, res, r.TLS.PeerCertificates[0])
+		}
+		glg.Infof("initVerifiers: added access token verifier having param: %+v", a.atpParams)
+		verifiers = append(verifiers, atVerifier)
+	}
+
+	if a.verifyRoleToken {
+		rtVerifier := func(r *http.Request, act, res string) error {
+			return a.VerifyRoleToken(r.Context(), r.Header.Get(a.rtHeader), act, res)
+		}
+		glg.Info("initVerifiers: added role token verifier")
+		verifiers = append(verifiers, rtVerifier)
+	}
+
+	if len(verifiers) < 1 {
+		return errors.New("error no verifiers")
+	}
+
+	// resize
+	a.verifiers = append([]verifier(nil), verifiers...)
```

Comment:
```suggestion
	copy(a.verifiers[0:],  verifiers)
```

Reply:
`a.verifiers` does not have an element.
So I can't use `copy()` as it is (nil is assigned to `a.verifiers`).
So I used make.  0c73eaa

Reply:
I think https://github.com/yahoojapan/athenz-authorizer/commit/0c73eaaa47d6d54a105fc7160cd7ead65d23af17 looks good but I read your first code, that was bad
```
a.verifiers = append([]verifier(nil), verifiers...)
```
why don't you allocate a.verifiers at L191 ? in this case we don't need to define verifiers

Reply:
understand, thank you.